### PR TITLE
1110772 - when changing basechannel the old childchannels are not selected by default

### PR DESCRIPTION
**Pull request by lneves12**
## What does this PR change?

This PR fixes a regression bug on the refactor of the subscription channels to React. When we select a new base channel a preselection should be done to all the similar child channels, so a new flag `wasSelectedOnOldBaseChannel` was added to the fetching child channels API.

Assumptions: The preselection is only applied on the first change, afterwards the user selection is always assumed.

I also added a new behavior that when a base channel is fetching the respective child channels all the other base channels selection is disabled to avoid weird bugs of inconsistent states.

I wonder if the old code about base channel assignment shouldn't be removed. It seems that it's not being used anymore. @mateiw is there any plan for that?

Note: I first implemented this logic in Javascript because I though the algorithm presented in the method `findCompatibleChildren` wasn't being used anymore, which is not true, thus keeping 2 code path wasn't  a good idea. I created a gist with it implemented in javascript just for reference, in case we need it later: https://gist.github.com/lneves12/bb8f0ad9e325aa7f19fa2a455be3b5fd

## GUI diff

Example 1:

Subscribed channels:
screenshot from 2018-11-15 10-02-05

After change base channel:
screenshot from 2018-11-15 10-02-18

Example 2:

Subscribed channels:
screenshot from 2018-11-15 09-59-13

After change base channel:
screenshot from 2018-11-15 10-00-17

- [x] **DONE**

## Documentation
- No documentation needed:

- [x] **DONE**

## Test coverage
- No tests.

- [x] **DONE**

## Links

Fixes #https://github.com/SUSE/spacewalk/issues/6018

- [x] **DONE**

### CODE REVIEW

**Comment by lneves12** [comment]
This PR isn't WIP anymore, so you can make a final review. Thanks!

File: branding/spacewalk-branding.changes [hunk 1]
```diff
@@ -1,3 +1,4 @@
+- Better label visualization when the input is disabled
```
Comment (mcalmer) [inline branding/spacewalk-branding.changes]: Please add the bugnumber to the changelog . For a submit we need at least one. So we should use every chance we have.

File: java/code/src/com/suse/manager/webui/controllers/SystemsController.java [hunk 2]
```diff
@@ -321,6 +323,17 @@ public static Object getAccessibleChannelChildren(Request request, Response resp
...
             }
+
+            Channel oldBaseChannel =  null;
+            try {
+                Long oldBaseChannelId = Long.parseLong(request.queryParams("oldBaseChannelId"));
+                oldBaseChannel = ChannelManager.lookupByIdAndUser(oldBaseChannelId, user);
+            }
+            catch (Exception e) {
+                //This is not a critical feature for this request, we want to keep move forward anyway
+                LOG.debug("An oldBaseChannelId parameter was provided when fetching childchannels");
```
Comment (mcalmer) [inline java/code/src/com/suse/manager/webui/controllers/SystemsController.java]: Is here a "not" missing? I do not understand the log message.

Comment (lneves12) [reply java/code/src/com/suse/manager/webui/controllers/SystemsController.java]: yes! A not was missing here, thanks.

I wonder if this LOG.debug shouldn't be an error :thinking: :thinking: 

If there is a bug on the frontend and this parameter oldBaseChannelId isn't provided, we will never be aware that something is wrong with the pre-selection. Should we consider this as critical and abort the request with an error instead of providing the child channels anyway with the pre selection info off?

File: branding/spacewalk-branding.changes (hunk 1 above)
Comment (lneves12) [reply branding/spacewalk-branding.changes]: Done

File: java/code/src/com/suse/manager/webui/controllers/SystemsController.java (hunk 2 above)
Comment (mcalmer) [reply java/code/src/com/suse/manager/webui/controllers/SystemsController.java]: Well, I am not sure if you hit this if the system does not have a channel. This can happen for Bootstrap entitled systems.
But "info" or "warn" would be good.
(Info is not logged by default, while warn is)

File: web/html/src/manager/products.js [hunk 3]
```diff
@@ -423,7 +423,7 @@ class Products extends React.Component {
   };
 
   searchData = (datum, criteria) => {
-    if (criteria) {
+    if (criteria && datum.label) {
```
Comment (ncounter) [inline web/html/src/manager/products.js]: Is this change related to this PR? It doesn't look like. If so, please remove it from here and add the fix (that makes sense) into a separate commit and PR. Thank you.

File: java/code/src/com/suse/manager/webui/controllers/SystemsController.java (hunk 2 above)
Comment (lneves12) [reply java/code/src/com/suse/manager/webui/controllers/SystemsController.java]: hum ok,
I changed it to log only if the query parameter oldBaseChannelId is provided and something unexpected happen

File: java/code/src/com/suse/manager/webui/utils/gson/ChannelsJson.java [hunk 4]
```diff
@@ -51,13 +53,16 @@
          * @param customIn custom channel flag
          * @param subscribableIn subscribable flag
          * @param recommendedIn the channel is recommended by its parent channel
+         * @param wasSelectedOnOldBaseChannelIn the channel was selected on the previous baseChannel
```
Comment (ncounter) [inline java/code/src/com/suse/manager/webui/utils/gson/ChannelsJson.java]: Why not something like `isCompatibleWithPreviousSelection` or `isCompatibleWithOldSelection`?

I know, as usual, naming things is something really hard, I already like your naming, but I am looking for something that explains the *feature intention* itself. Reading `wasSelectedOnOldBaseChannelIn` could mislead the fact that exactly the same child channel was selected already before, and this is not the case since we are pre-selecting the **correspondent** child channel under the new base channel, which is different from the previous one (`channel id` is different for instance). It's just a hint to the future developer who will need to take a look here. Sorry for being picky :wink:

File: java/code/src/com/suse/manager/webui/controllers/SystemsController.java [hunk 5]
```diff
@@ -334,6 +351,9 @@ public static Object getAccessibleChannelChildren(Request request, Response resp
                             ResultJson.error("not_a_base_channel"));
                 }
 
+                Collection<Channel> preservations =
```
Comment (ncounter) [inline java/code/src/com/suse/manager/webui/controllers/SystemsController.java]: This naming I like it a lot :+1:

File: java/code/src/com/suse/manager/webui/utils/gson/ChannelsJson.java (hunk 4 above)
Comment (lneves12) [reply java/code/src/com/suse/manager/webui/utils/gson/ChannelsJson.java]: yeah, isCompatibleWithPreviousSelection seems better, I will rename it

File: web/html/src/manager/products.js (hunk 3 above)
Comment (lneves12) [reply web/html/src/manager/products.js]: Sure, I will do it :)

**Comment by mcalmer** [comment]
from my part it is ok. There is a small checkstyle error which should be fixed.
